### DISCUSSION TITLE
Validation and validation of a MapAttribute and its properties

### DISCUSSION THREAD

**Discussion by dls314** [tywalch/electrodb#432] [Q&A]
Hi,

I have a couple of questions about validation.

* Are validate functions called for non-required attributes or properties that are undefined?
  * I think the answer is no

Also, a question about how validation runs for a Schema's MapAttribute when that MapAttribute has its own top level validate function as well as validate functions at its properties.

* Is a validate function on a MapAttribute property called before or after the MapAttribute's own top level validate function?
  * I haven't been able to figure this out yet, but I suspect the top level validate function is called after property level validate functions

Thanks!

**Comment by tywalch** [comment]
Hi @dls314 👋 

Great question! I am not currently on my computer, but wanted to at least reply; let me know if you'd like more detail and I can write out some examples when I'm back at my desk. 

Validation is applied to the attribute or nested attribute of the value being *set*. For example, if you set a map attribute with nested attributes, validation will be ran on those nested attributes first, and then lastly the map attribute. Alternatively, if you set a nested attribute directly, it will only validations for that attribute and not its map. 

The reasoning for this is because the values not being set are not known. In the second example above (where a nested attribute is set) ElectroDB only has the value of the nested attribute provided. If the containing map had a validation checking for values not being set, that function would likely throw because they would appear missing. 

Hope this helps, but let me know 👍

**Comment by dls314** [comment]
Thanks! That was exactly what I needed

**Comment by dls314** [comment]
Hi @tywalch,

I have a related question in this area about how a `MapAttribute` works with validation.

When `MapAttribute`'s base `Attribute`'s `isValid` method from https://github.com/tywalch/electrodb/blob/master/src/schema.js#L618-L631 runs, it calls to `MapAttribute`'s `_isType` from https://github.com/tywalch/electrodb/blob/master/src/schema.js#L747-L778 and if the `value` argument to `_isType` is undefined, and the attribute is not required, then the return is `[true, []]` at https://github.com/tywalch/electrodb/blob/master/src/schema.js#L758.

From that return, `isValid` sets `isTyped` to true at https://github.com/tywalch/electrodb/blob/master/src/schema.js#L620, which results in running calling `Attribute`'s `validate` function from https://github.com/tywalch/electrodb/blob/master/src/schema.js#L622 (which was set to a wrapper around the attribute definition's `validate` function if provided)

This matches up with what I'm seeing in my usage, where my validate function is being called for a non-required undefined property while trying to create an item.

It seems wrong that validation is being called for a non-required, undefined property.

Did I miss something while reading through this?

**Reply by dls314** [reply]
It seems like the various `_isType` methods should return `[this.required, reason]` at

https://github.com/tywalch/electrodb/blob/master/src/schema.js#L566
https://github.com/tywalch/electrodb/blob/master/src/schema.js#L758
https://github.com/tywalch/electrodb/blob/master/src/schema.js#L953
https://github.com/tywalch/electrodb/blob/master/src/schema.js#L1151
